libtcmu.c

```diff
@@ -173,6 +197,433 @@ static struct tcmulib_handler *find_handler(struct tcmulib_context *ctx,
+
+	/* cleanup push/pop not _really_ required here atm */
+	pthread_cleanup_push(_cleanup_mutex_lock, &io_wq->io_lock);
+	pthread_mutex_lock(&io_wq->io_lock);
+
+	list_add_tail(&io_wq->io_queue, &io_entry->entry);
+	pthread_cond_signal(&io_wq->io_cond); // TODO: conditional
+
+	pthread_mutex_unlock(&io_wq->io_lock);
+	pthread_cleanup_pop(0);
+
+	return TCMU_ASYNC_HANDLED;
+}
+
+static int setup_io_work_queue(struct tcmu_device *dev)
+{
+	int ret;
+	struct tcmu_io_queue *io_wq = &dev->work_queue;
+
+	list_head_init(&io_wq->io_queue);
+
+	ret = pthread_mutex_init(&io_wq->io_lock, NULL);
+	if (ret < 0) {
+		goto out;
+	}
+	ret = pthread_cond_init(&io_wq->io_cond, NULL);
+	if (ret < 0) {
+		goto cleanup_lock;
+	}
+
+	// TODO: >1 worker threads (per device via config)
+	ret = pthread_create(&io_wq->io_wq_thread, NULL, io_work_queue, dev);
+	if (ret < 0) {
+		goto cleanup_cond;
+	}
+
+	return 0;
+
+cleanup_cond:
+	pthread_cond_destroy(&io_wq->io_cond);
+cleanup_lock:
+	pthread_mutex_destroy(&io_wq->io_lock);
+out:
+	return ret;
+}
+
+static void cleanup_io_work_queue(struct tcmu_device *dev,
+				  bool cancel)
+{
+	int ret;
+	struct tcmu_io_entry *curr, *next;
+	struct tcmu_io_queue *io_wq = &dev->work_queue;
+
+	if (cancel) {
+		cancel_thread(io_wq->io_wq_thread);
+	}
+
+	list_for_each_safe(&io_wq->io_queue, curr, next, entry) {
+		list_del(&curr->entry);
+		free(curr);
+	}
+
+	ret = pthread_mutex_destroy(&io_wq->io_lock);
+	if (ret != 0) {
+		tcmu_err("failed to destroy io workqueue lock\n");
+	}
+
+	ret = pthread_cond_destroy(&io_wq->io_cond);
+	if (ret != 0) {
+		tcmu_err("failed to destroy io workqueue cond\n");
+	}
+}
+
+static int setup_aio_tracking(struct tcmu_device *dev)
+{
+	int ret;
+	struct tcmu_track_aio *aio_track = &dev->track_queue;
+
+	aio_track->tracked_aio_ops = 0;
+	ret = pthread_spin_init(&aio_track->track_lock, 0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static void cleanup_aio_tracking(struct tcmu_device *dev)
+{
+	int ret;
+	struct tcmu_track_aio *aio_track = &dev->track_queue;
+
+        assert(aio_track->tracked_aio_ops == 0);
+
+	ret = pthread_spin_destroy(&aio_track->track_lock);
+	if (ret < 0) {
+		tcmu_err("failes to destroy track lock\n");
+	}
+}
+
+static void async_call_command(struct tcmu_device *dev,
+			       struct tcmulib_cmd *cmd)
+{
+	int ret;
+	struct tcmulib_handler *handler = tcmu_get_dev_handler(dev);
+	struct tcmur_handler *r_handler = handler->hm_private;
+
+	if (r_handler->aio_supported) {
+		ret = invokecmd(dev, cmd);
+	} else {
+		ret = aio_schedule(dev, cmd);
+	}
+
+	/*
+	 * command (processing) completion is done when one of the
+	 * following scenario occurs:
+	 *  - synchronous handler:
+	 *	only if aio_schedule() returns an error
+	 *  - asynchronous handler:
+	 *	on an error or if the command was not asynchronously
+	 *	handled (see generic_handle_cmd(), non store callouts)
+	 */
+	if (ret != TCMU_ASYNC_HANDLED) {
+		tcmulib_command_complete(dev, cmd, ret);
```

Comment:
On failure, aio_schedule needs to return a SAM_STAT value from scsi_defs.h. For ENOMEM type of errors I think SAM_STAT_TASK_SET_FULL is similar to what the kernel lio code uses.